### PR TITLE
Test framework

### PR DESCRIPTION
**Pull request by aljones15**
1. Adds eslint as a dependency to bedrock-account-http
2. Expands Test project with tests for accountExists
3. Adds Tests for account create
4. Adds Tests for account get by id
5. Adds a stub that let's tests specify the req.user the handler receives
6. abstracts lambdas out to async functions with doc string with tests to ensure they still work.
7. adds a check to account create for when the user forgets to specify an email. (ui check is there).

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -39,44 +39,75 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
+   * @description
+   */
+  async function createNewAccount(req, res) {
+    const account = {
+      id: 'urn:uuid:' + uuid(),
+      email: req.body.email
+    };
+    if(!account.email) {
+      throw new BedrockError(
+        'No Email', 'InvalidEmailError', {
```
Comment (dlongley) [inline lib/index.js]: We try to keep a very minimal set of "error types" (and they typically mirror those [found here](https://developer.mozilla.org/en-US/docs/Web/API/DOMException#Error_names)). That being said, we shouldn't have to check for `email` here -- as it should be done through the validator. It looks like we never got around to implementing the validator, though. So it looks like that's what needs to be done, see the TODO below:

```
// TODO: validate('services.account.postAccounts')
```

File: lib/index.js [hunk 2]
```diff
@@ -39,44 +39,75 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
+          public: true,
+        });
+    }
+    const meta = {
+      sysResourceRole: [{
+        sysRole: 'account.registered',
+        resource: [account.id]
+      }]
+    };
+      // actor is not required, but passed through if given
```
Comment (dlongley) [inline lib/index.js]: Indent is off on the comment

File: lib/index.js [hunk 3]
```diff
@@ -39,44 +39,75 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
-        }]
-      };
-      // actor is not required, but passed through if given
-      const {actor = null} = req.user || {};
-      await brAccount.insert({actor, account, meta});
-      res.status(201).location(baseUri + '/' + account.id).json(account);
-    }));
-
+    asyncHandler(createNewAccount));
+  /**
```
Comment (dlongley) [inline lib/index.js]: Needs empty line of whitespace to break this up (though the function below is a private helper so it should move to the bottom of the file anyway).

File: lib/index.js [hunk 4]
```diff
@@ -39,44 +39,75 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
   const accountPath = basePath + '/:account';
   const {baseUri} = bedrock.config.server;
 
+  /**
+   * @function createNewAccount
+   * @param {Request} req
+   * @param {Response} res
+   * @description
+   */
+  async function createNewAccount(req, res) {
```
[the remark this answers is not in the document]
Comment (aljones15) [reply lib/index.js]: ok so basically optionalAuth stuff is first or you mean move all the functions to the bottom and rely on hoisting?

File: lib/index.js (hunk 1 above)
Comment (aljones15) [reply lib/index.js]: I can switch to DataError instead

File: lib/index.js (hunk 2 above)
Comment (aljones15) [reply lib/index.js]: weird ok looking into it.

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: we could actually separate the actual handlers into their own module and then call this routes we have yet to implement delete and patch for instance so doing it now would mean when those handlers are added we expand in another file. separate routes from handler.

File: lib/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlongley) [reply lib/index.js]: Move all to the bottom and rely on hoisting. The goal is to have it be such that a dev reading the file will bump into the stuff they are most interested in first (public API stuff).

File: lib/index.js (hunk 1 above)
Comment (dlongley) [reply lib/index.js]: I would normally say that's the right thing to do here -- but the validator will take care of it. So we should implement the validator instead and remove the special code.

File: lib/index.js (hunk 3 above)
Comment (dlongley) [reply lib/index.js]: Own module or own file?

Comment (dlongley) [reply lib/index.js]: I just want to make sure we've got the right organizational balance here ... if we're creating too few files it's a mess, if we're creating too many it's too hard to navigate/compare/find what you want. We certainly don't want another module... but a better file structure might be ok.

cc: @davidlehn, @mattcollier

**Comment by aljones15** [comment]
Regarding the function names:

1. named functions will show up in a stack trace when an error occurs making debugging easier.
2. take getAccountById in order to test we have stub bedrock-passport's optional auth if we did this:

```
export async function getAccountById(res,req) {
  const {actor} = req.user || {}
}

it('should return an account', function() {
  const req = {user: AdminUser, params: {account: testAccount}}
  const res = new MockResponse();
  const result = await getAccountById(req, res);
})

```

no need for a stub we can now pass in a req.user which can be configured easily to test multiple different auth scenarios quickly. Additionally, this makes testing validators etc. easier too.

3. lambdas are really intended for small stuff and on call the function is created then executed. Named functions are merely executed so small performance increase.

4. modularity we can now separate routes from handlers meaning that handlers can stored in a separate module resulting in cleaner code.

5. easier documentation.

so basically instead of getting a line number in an error report we get an actual function name, the ability to test the handlers easier, and a slight performance increase.

File: lib/index.js (hunk 1 above)
Comment (aljones15) [reply lib/index.js]: that is fairly easy to do where should I put the validator? in the file?

1. implement validator object as middle ware that calls next with a DataError
2. validate is a switch

Comment (mattcollier) [reply lib/index.js]: @aljones15 we have a module for that :)  `bedrock-validation` which among other things returns an express middleware that is used as shown in some of the commented code.  I'll share more with you about this in another context.

**Comment by mattcollier** [comment]
@aljones15 here's some details about how to implement validation: https://gist.github.com/mattcollier/bdc16c556ecd31fdab0f6157ca5ae5c7

File: lib/index.js (hunk 1 above)
Comment (aljones15) [reply lib/index.js]: ok cool I will look at bedrock-validation and add this to it. thanks man.

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: I am actually am voting for a separate file that we import from. looking at the file it's just kind of complicated. we have several event emitter callbacks and then several different middleware compositions. if we move the handler functions to a different file it means routes will clarify which middleware goes where etc. and handlers can be found more easily.

**Comment by dlongley** [comment]
@aljones15,

> lambdas are really intended for small stuff and on call the function is created then executed. Named functions are merely executed so small performance increase.

Ignoring the other points for the moment -- just wanted to say that this performance increase does not apply here because the arrow function is not recreated on every call.

File: lib/config.js [hunk 5]
```diff
@@ -12,6 +12,6 @@ cfg.routes = {
 };
 
 // TODO: implement validation
```
Comment (mattcollier) [inline lib/config.js]: Remove todo :)

File: schemas/account.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const email = {
+  type: 'string'
```
Comment (mattcollier) [inline schemas/account.js]: We have stronger schema here: https://github.com/digitalbazaar/bedrock-validation/blob/master/schemas/email.js

`schemas` are exported in bedrock-validation here: https://github.com/digitalbazaar/bedrock-validation/blob/master/lib/index.js#L21

They are implemented as schema factories, so you do `schema.email()`.

File: test/mocha/10-api.js [hunk 7]
```diff
@@ -32,14 +32,10 @@ describe('bedrock-account-http', function bedrockAccountHttp() {
...
+    it('should return 400 with no email', async function worksGreat() {
       const result = await api.get('/');
-      const {status, data} = result;
-      should.equal(status, 404);
-      data.should.be.an('object');
-      const {message, type} = data;
-      message.should.match(/account does not exist/i);
-      type.should.match(/NotFoundError/i);
+      const {status} = result;
+      should.equal(status, 400);
```
Comment (mattcollier) [inline test/mocha/10-api.js]: Validation middleware does return an error in response.body, so you *could* make some light asserations about that.

File: schemas/account.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+};
+
+const postAccounts = {
+  title: 'bedrock-accounts-http account creation post',
```
Comment (mattcollier) [inline schemas/account.js]: Add: `additionalProperites: false` where appropriate, otherwise, the default is that other properties are allowed.

https://json-schema.org/understanding-json-schema/reference/object.html

File: schemas/account.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+const email = {
+  type: 'string'
+};
+
+const postAccounts = {
```
Comment (mattcollier) [inline schemas/account.js]: So, this type of validator would look something like this:
```
{
  title: 'foo',
  type: 'object',
  required: ['foo', 'bar'],
  additionalProperties: false,
  properties: {
    foo: {...},
    bar: {...},
  }
}
```

File: test/mock.data.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+/* jshint node: true */
```
Comment (mattcollier) [inline test/mock.data.js]: I know this was cut/paste, but we don't need this.

File: lib/index.js [hunk 11]
```diff
@@ -34,6 +35,7 @@ bedrock.events.on('bedrock-session-http.session.get', (req, session) => {
 });
 
 bedrock.events.on('bedrock-express.configure.routes', app => {
+  app.use(boolParser());
```
Comment (dlongley) [inline lib/index.js]: This will update the whole express app -- so not the right way to do it.

@davidlehn -- can you look at this and suggest the proper way?

Comment (davidlehn) [reply lib/index.js]: It looks to be just regular middleware.  This could just be added to endpoints that need it.  Unclear what the ramifications of adding it everywhere would be.

File: test/mocha/10-api.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+const baseURL =
+ `https://${config.server.host}${config['account-http'].routes.basePath}`;
+
+// simple quick func to check validation errors
+// TODO extend mocha should with this
+function validationError(result, errorMethod, expectedStatus = 400) {
+  result.status.should.equal(expectedStatus);
+  result.data.should.have.property('message');
+  result.data.should.have.property('type');
```
Comment (mattcollier) [inline test/mocha/10-api.js]: At minimum I think we want to make assertion about `type` value.

Comment (aljones15) [reply test/mocha/10-api.js]: ok cool doing it. Also we could make this a custom should method on chai, but that might take some time to do (I have done custom expect, but not should).

File: lib/index.js (hunk 3 above)
Comment (aljones15) [reply lib/index.js]: this the only remaining task so I will let it say at the bottom of the current module for now.

**Comment by davidlehn** [comment]
Regarding extracting the main route handler functions:
- I think it makes the code less readable and maintainable. When you want to work on an endpoint you have to find the route, then find a function. And putting logic in other files would make this process even more difficult.
- If the handlers are getting too large, then look at why.  In most cases we move code out into libraries as needed.  Seems like too much overhead to move all code away from the route site.
- Documentation of the endpoints themselves seems like enough to me.  (We need lots more work on this)  Seems like it would duplicate effort to re-document an internal function that does all the work.
- Having anonymous functions seems fine.  It's never been a problem so far.  Line numbers are fine to find where to debug.  (Wait... why are people adding bugs in the first place?!)
- Is the testing issue real?  We have (I think) had testing of logged in users and anonymous users and similar.  Perhaps we can just improve on that pattern and make some helper utilities if needed?
- This seems like a lot of overhead for unknown gain, and at least some pain.
- And the biggest problem is that none of our other code does this.  IMHO, we should not just switch over one module like this.  If anything it should be made more consistent with our other code.
- As you can tell, I'm not in favor of that part of this patch.

If we decide we want to explore other code layout patterns, I suggest making new branches specifically to flesh out the ideas.  Mixing that in with other changes seems like a bad idea.

**Comment by davidlehn** [comment]
Please track high level developer facing changes in the changelog as you go.  I wrote a quick note about this (needs work) in the bedrock contributing guide. Lots of examples of style and level of detail in every other module.
https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md#changelogs

**Comment by aljones15** [comment]
> * Documentation of the endpoints themselves seems like enough to me.  (We need lots more work on this)  Seems like it would duplicate effort to re-document an internal function that does all the work.

> * Having anonymous functions seems fine.  It's never been a problem so far.  Line numbers are fine to find where to debug.  (Wait... why are people adding bugs in the first place?!)

would adding named function inside the handler be ok?

```
  app.post(
    basePath,
    optionallyAuthenticated,
    validate('account.postAccounts'),
    asyncHandler(
      async function createNewAccount(req, res) {
        /**
         * @param {Request} req
         * @param {Response} res
         * @description
        */
    const account = {
      id: 'urn:uuid:' + uuid(),
      email: req.body.email
    };
    const meta = {
      sysResourceRole: [{
        sysRole: 'account.registered',
        resource: [account.id]
      }]
    };
    // actor is not required, but passed through if given
    const {actor = null} = req.user || {};
    await brAccount.insert({actor, account, meta});
    res.status(201).location(baseUri + '/' + account.id).json(account);
  }));
```

> * Is the testing issue real?  We have (I think) had testing of logged in users and anonymous users and similar.  Perhaps we can just improve on that pattern and make some helper utilities if needed?

it is real and was solved using sinon,

**Comment by davidlehn** [comment]
> > * Documentation of the endpoints themselves seems like enough to me.  (We need lots more work on this)  Seems like it would duplicate effort to re-document an internal function that does all the work.
> 
> why not just use the handler doc string for the end point?
> how are you proposing documenting the endpoints? js-doc has a lot of
> ways of doing it, but documenting the functions used by the routes seemed the easiest to me.
> So can we agree on an way of documentating end points preferably through javascript comments and I will start using it today.
> 

The http api and js api docs will be different and are not in general targeting the same audience.  We have this:
https://github.com/digitalbazaar/bedrock-docs
Which uses RAML and the docs load from a live server.  I think plan was to be able to extract docs out to a static site too(?).  It all needs work, may be bit rotten at the moment, but we have used it in some places.  Honestly, so much is a moving target and has only had internal use so time and resources on such docs has been minimal.

I don't know what we want to do about all this and it really needs input from everyone.  It's another thing I'd suggest be put in a branch for discussion rather than mixing in with other commits.


> > * Having anonymous functions seems fine.  It's never been a problem so far.  Line numbers are fine to find where to debug.  (Wait... why are people adding bugs in the first place?!)
> 
> would adding named function inside the route be ok?
> 
> ```js
>   app.post(
>     basePath,
>     optionallyAuthenticated,
>     validate('account.postAccounts'),
>     asyncHandler(
>       async function createNewAccount(req, res) {
>         /**
>          * @param {Request} req
>          * @param {Response} res
>          * @description
>         */
> ...
> ```
> ...

What would req and res docs say?  I think that is better handled as http api docs.

> > * Is the testing issue real?  We have (I think) had testing of logged in users and anonymous users and similar.  Perhaps we can just improve on that pattern and make some helper utilities if needed?
> 
> it is real and was solved using sinon. It's not that big of a deal though because bedrock-account already handles unit testing the functions the handler composes however using route and handler separate does let us test handlers with out middleware.
> 

My intuition is saying that's a specialized case vs a general one?  How many handlers need testing without middleware?


> I think the best pattern is
> 
> ```js
> async function handlerCreateAccount(req, res) {
> 
> }
> // this is after the function delcaration
> app.post(
>   routeAddress,
>   ensureAuthenticationMiddleware,
>   boolParser(),
>   validator('account.hasEmail'),
>   asyncHandler(handlerCreateAccount))
> 
> // this is at the bottom of the file
> module.exports.createAccount = handlerCreateAccount;
> ```
> 
> this was we have both things fairly close each other with the ability to test with out middleware if needed.

I guess that's better.  I'm not sure.  We have to test the http api anyway, so it seems that, in general, those tests would/should cover whatever the function tests would do.  It's difficult to know what all the middleware here and globally will do so we can't only test the handler functions.

I'll mention again, I think these major code style updates should be on their own branch so we can flesh out ideas.  I think it's better to just leave things using the old patterns so we're consistent across all our repos until we decide to change the style going forward.  I mostly understand why these changes may be a good idea, it's just difficult to jump into the middle of a test/validation PR with company wide style discussion that really requires everyone to participate.

**Comment by aljones15** [comment]
> The http api and js api docs will be different and are not in general targeting the same audience. We have this:
> https://github.com/digitalbazaar/bedrock-docs
> Which uses RAML and the docs load from a live server. I think plan was to be able to extract docs out to a static site too(?). It all needs work, may be bit rotten at the moment, but we have used it in some places. Honestly, so much is a moving target and has only had internal use so time and resources on such docs has been minimal.

@davidlehn Let's do this in a separate PR.

**Comment by davidlehn** [comment]
Docs in another PR sounds good.  Though I notice there are odd jsdocs in there now that are describing the params to `app.VERB()` calls?  Should those be part of a doc PR?

A style question, are the latest eslint rules forcing this:
```js
  }
  )
```
vs single line closing:
```js
  })
```
I'm still not convinced the named async functions are needed over the more succinct anonymous arrow functions.  Newer VMs are much better with async/await stack traces than they used to be.  I haven't compared the output recently though.

We seem to be almost back where we started. ;-)

File: lib/index.js [hunk 13]
```diff
@@ -39,11 +40,19 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
+   * @param {Function} handler
+   * @description
+   */
   app.post(
     basePath,
     optionallyAuthenticated,
-    // TODO: validate('services.account.postAccounts')
-    asyncHandler(async (req, res) => {
+    validate('account.postAccounts'),
+    asyncHandler(async function createNewAccount(req, res) {
```
Comment (gannan08) [inline lib/index.js]: For consistency purposes and developer friendliness, I prefer the arrow function pattern in our express routes.

File: package.json [hunk 14]
```diff
@@ -23,16 +23,21 @@
   "dependencies": {
     "cors": "^2.8.4",
     "express-async-handler": "^1.1.3",
+    "express-query-boolean": "^2.0.0",
```
Comment (gannan08) [inline package.json]: Is this package essential? I believe the bedrock-account-http package would function sufficiently without the need of converting strings to booleans, we should be weary pulling in outside packages that are not necessary.

Comment (mattcollier) [reply package.json]: @gannan08 I believe this is a good solution.  The issue is that a query parameter `?foo=true` shows up in express as a string, and this throws a wrench in the works when one would like to validate the query parameter as a boolean using json-schema/bedrock-validation.  

Perhaps the *best* solution is to have our own implementation of this included in bedrock-express, but these things take time and haggling.

Comment (gannan08) [reply package.json]: > this throws a wrench in the works when one would like to validate the query parameter as a boolean using json-schema/bedrock-validation.

Okay, so we convert expected parameters to booleans to check against our validator. That is cool with me.

> Perhaps the _best_ solution is to have our own implementation of this included in bedrock-express, but these things take time and haggling.

Agreed, we're definitely not at a point where we can spend time on haggling things like this. Don't know if we want to file an issue to come back to this later or add a FIXME in the code.

Comment (aljones15) [reply package.json]: added a FIXME to the route and require.

File: lib/index.js [hunk 15]
```diff
@@ -63,19 +65,31 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
+        throw new BedrockError(
+          'Account does not exist.', 'NotFoundError', {
+            httpStatusCode: 404,
+            public: true,
+          });
       }
       // TODO: call `brAccount.getAll`
-      next();
+      // TODO: improve error details
+      throw new BedrockError(
```
Comment (dlongley) [inline lib/index.js]: This seems odd to change this to an account not found error ... but since we're about to implement adding `getAll` here I suppose it doesn't matter.

File: lib/index.js [hunk 16]
```diff
@@ -96,7 +111,9 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
       const {actor} = req.user || {};
       const record = await brAccount.get({actor, id});
       res.status(200).json(record);
-    }));
+    }
+    )
+  );
```
Comment (dlongley) [inline lib/index.js]: This is some funky stuff... do we need to fix the eslint rules so it doesn't happen?

File: test/helpers.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+api.createAccount = email => {
+  const newAccount = {
+    id: 'urn:uuid:' + uuid(),
+    email
+  };
+  return newAccount;
+};
+
+api.createIdentity = userName => {
```
Comment (dlongley) [inline test/helpers.js]: I don't think this is used anywhere ... can we remove? Same for anything else unused.

File: lib/index.js (hunk 15 above)
Comment (aljones15) [reply lib/index.js]: incorrect actually or maybe I'm wrong, but I actually implemented the getAll on an admin route I did not re-use this route, but it's less secure.

Comment (aljones15) [reply lib/index.js]: anyways this is more of an issue for the next PR i.e. the get all accounts by email, because we might want to re-use this route with a different query string rather than create a new one.

File: lib/index.js (hunk 16 above)
Comment (aljones15) [reply lib/index.js]: fixed! will look into it a linter issue for this in a moment.
